lkl: 32-bit timespec fixes

ddiss opened the pull request:
```
The following changes since commit b357b84371983c7de658fec83265b9fe7e48c20a:

  Merge pull request #555 from ddiss/lkl_test_icmp_buf_overflow (2025-01-29 17:36:39 -0800)

are available in the Git repository at:

  https://github.com/ddiss/linux lkl_tests_parallel_build_fix

for you to fetch changes up to 0692c1e05031ac61758d22a14417b50ab81732e2:

  lkl: fix 32-bit timespec casts (2025-02-06 09:47:30 +1100)

----------------------------------------------------------------
David Disseldorp (5):
      lkl tools: use 64-bit lkl_sys_nanosleep on 32-bit builds
      lkl tools: fix some -Wincompatible-pointer-types errors
      lkl cptofs: handle stat errors earlier
      lkl: rework broken lkl_sys_select timeval conversion
      lkl: fix 32-bit timespec casts

 tools/lkl/cptofs.c         | 44 ++++++++++++++++++++++++++++----------------
 tools/lkl/include/lkl.h    | 36 +++++++++++++++++-------------------
 tools/lkl/lib/fs.c         |  5 ++---
 tools/lkl/lib/posix-host.c | 21 ++++++++++++++++++---
 tools/lkl/lklfuse.c        | 20 ++++++++------------
 tools/lkl/tests/boot.c     |  7 ++++---
 tools/lkl/tests/net-test.c |  4 ++--
 7 files changed, 79 insertions(+), 58 deletions(-)
```

ddiss commented:
Regarding the checkpatch warnings: two are from unwrapped compiler output in the commit messages and one is for a `/* fall through */` comment in code copied from `test.c` -> `test.h`. I'd prefer to leave them as-is but can change if required.

I have some more 32-bit time_t fixes which I'll push here soon.

ddiss commented on tools/lkl/tests/Build:
```diff
@@ -1,6 +1,6 @@
-boot-y += boot.o test.o
```
I hit the build failure quite consistently on the openSUSE Build Service: https://build.opensuse.org/package/show/home:ddiss:lkl/lkl . I builds with `make -O -j4 V=1 VERBOSE=1 -C tools/lkl KCONFIG=opensuse_defconfig` (make-4.4.1-3.2).

The above project link is patched with this fix. The c221333820c57a4520dca307680ceb1340a89350 commit message carries truncated output.

tavip replied:
I was able to reproduce the issue after loading my system with cpu/io workloads. The issue seems to be that the tools build system can only handle one object when building and because we build in parallel multiple the %-in.o rules that in turns starts multiple builds for test.o and cla.o. 

I don't like this solution as the problem is not only limited to tests, it will happen when .o files are shared between targets.  Can you please carry this patch in your tree, while I work on a fix? Created #558 to track it.

ddiss replied:
Sure, happy to carry this downstream for now. I didn't see any other shared .o files, as the OBS tends to only trip over this in tests.

thehajime commented on tools/lkl/include/lkl.h:
```diff
@@ -64,13 +64,12 @@ static inline int lkl_sys_fstatfs(unsigned int fd, struct lkl_statfs *buf)
-static inline int lkl_sys_nanosleep_time32(struct lkl_timespec *rqtp,
-					   struct lkl_timespec *rmtp)
+static inline int lkl_sys_nanosleep(struct __lkl__kernel_timespec *rqtp,
+				    struct __lkl__kernel_timespec *rmtp)
```
> Then we can remove `__lkl_kernel_timespec` from LKL libs and apps which we introduced in [3d4047a](https://github.com/lkl/linux/commit/3d4047ac9aae49baa2c7cedb02304a1e5ab33966). @thehajime does this sound reasonable?

sorry, I don't quite remember what/why my followup fixes (3d4047ac9aae49baa2c7cedb02304a1e5ab33966) did.. my best guess is to silence warnings from compilation.

tavip replied:
> Do you just want this wrapper for nanosleep, or should I also add lkl_timespec wrappers for lkl_sys_utimensat()? I was hoping to avoid going too far down the generic-libc path and just stick to the modern kernel API.

I thought more about this and I think you are right, we should stick with modern kernel APIs. 

We should be able to override `struct timespec`  defined in `include/uapi/linux/time.h` with a custom one `arch/lkl/include/uapi/asm/types.h`. If you prefer to merge as is  (use __lkl_kernel_timespec) for now we can cleanup later.

ddiss replied:
...
> We should be able to override `struct timespec` defined in `include/uapi/linux/time.h` with a custom one `arch/lkl/include/uapi/asm/types.h`. If you prefer to merge as is (use __lkl_kernel_timespec) for now we can cleanup later.

I'm not sure I follow; would you like some sort of helper macro to ease conversion from `timespec` to `__lkl_kernel_timespec`? Thanks a lot for the review btw

tavip replied:
Lets merge this as it is and I will followup with a proposal.

ddiss commented:
I've reworked the `lkl_sys_select timeval conversion` fix to drop the overflow detection, as discussed with @thehajime in https://github.com/lkl/linux/issues/557

ddiss commented:
Changes since previous version:
- drop c221333820c5 ("lkl tests: fix parallel build failures") now tracked via https://github.com/lkl/linux/issues/558

tavip reviewed (approved):
LGTM, great work @ddis, thank you!
